Make FakeProcessGroup traceable

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113314

This PR mimics what we have done to trace ProcessGroup. This allows use to use FakeProcessGroup with torch.compile. FakeProcessGroup allows us to use world_size > 1 without creating multiple processes thus enabling the usage of PDB to debug bucketing DDP allreduce in the Inductor. We can theoretically use GLOO with world_size==1 to achieve the same goal. However, the `wait()` seems to be optimized away when the world_size is 1.

Differential Revision: [D51136463](https://our.internmc.facebook.com/intern/diff/D51136463/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

fegin committed:
Make FakeProcessGroup traceable

This PR mimics what we have done to trace ProcessGroup.

Differential Revision: [D51136463](https://our.internmc.facebook.com/intern/diff/D51136463/)

[ghstack-poisoned]

wanchaol commented on torch/_dynamo/variables/builder.py:
```diff
@@ -660,6 +661,12 @@ def index_source(key):
                 source=self.source,
                 guards=self.make_guards(GuardBuilder.ID_MATCH),
             )
+        elif FakeProcessGroupVariable.is_process_group(value):
```
could we make this check directly inside `ProcessGroupVariable`, I think FakePG is just another type of PG so we should try to make that works directly

fegin replied:
Good point. I guess we won't actually do something special for FakePG. Let me change it.

fegin committed:
Update on "Make FakeProcessGroup traceable"


This PR mimics what we have done to trace ProcessGroup. This allows use to use FakeProcessGroup with torch.compile. FakeProcessGroup allows us to use world_size > 1 without creating multiple processes thus enabling the usage of PDB to debug bucketing DDP allreduce in the Inductor. We can theoretically use GLOO with world_size==1 to achieve the same goal. However, the `wait()` seems to be optimized away when the world_size is 1.

Differential Revision: [D51136463](https://our.internmc.facebook.com/intern/diff/D51136463/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm
